src/lib.rs

```diff
@@ -44,19 +49,27 @@ pub fn run_and_output(config: Config) {
 const MAX_FIX_ITERATIONS: usize = 10;
 
 pub fn run(config: Config) -> Vec<ViolationWithContext> {
-    let resolved_rules = config.get_resolved_rules();
-    let aggregated_queries = AggregatedQueries::new(&resolved_rules, &config);
+    let instantiated_rules = config.get_instantiated_rules();
+    let language = SupportedLanguage::Rust;
+    let aggregated_queries = AggregatedQueries::new(&instantiated_rules, language);
     let tree_sitter_grep_args = get_tree_sitter_grep_args(&aggregated_queries);
     let all_violations: Mutex<HashMap<PathBuf, Vec<ViolationWithContext>>> = Default::default();
     let files_with_fixes: AllPendingFixes = Default::default();
+    let instantiated_per_file_rules: AllInstantiatedPerFileRules = Default::default();
```

Comment:
Ok so we need to instantiate each rule "per-file run" but we don't know what those files are up front

So we need to lazily create per-file per-rule instances as we're running these `tree_sitter_grep` callbacks

It seemed to me like it would be too "high-contention" to put that behind one shared `Mutex`

So what I came up with was a "set of n hash-bucket `Mutex`'s" where the thing being hashed is the file path

I believe hashing on the file path (vs the rule name) is the right choice because `tree_sitter_grep` will only ever be running one query at a time for a given file (right? It only takes one query, so I don't think there's any "sub-file" parallelization happening) whereas two different `tree_sitter_grep` threads could happen to be invoking a callback for the same rule at the same time (for two different files)

So therefore in theory there shouldn't be contention if we hash by file path and there's enough buckets that multiple paths won't usually hash to the same bucket?

Reply:
Or use a concurrent hash map!

Reply:
This ended up landing somewhere where it wasn't necessary at all (by making it so that each file run has a "defined lifetime" within a single thread)

But ya a concurrent hashmap would again probably have made sense, I'm not used to thinking to "reach for it"